// openapi-diff/dist/openapi-diff/spec-differ/diff-finder/find-diffs-in-responses.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const get_added_keys_from_objects_1 = require("./common/get-added-keys-from-objects");
const get_removed_keys_from_objects_1 = require("./common/get-removed-keys-from-objects");
const create_difference_1 = require("./create-difference");
const findAddedDifferencesInResponses = (sourceResponses, destinationResponses) => {
    const addedResponses = get_added_keys_from_objects_1.getAddedKeysFromObjects(sourceResponses, destinationResponses);
    return addedResponses.map((statusCode) => {
        const parsedResponse = destinationResponses[statusCode];
        return create_difference_1.createDifference({
            action: 'add',
            destinationSpecOrigins: [parsedResponse.originalValue],
            propertyName: 'response.status-code',
            source: 'openapi-diff',
            sourceSpecOrigins: []
        });
    });
};
const findRemovedDifferencesInResponses = (sourceResponses, destinationResponses) => {
    const removedResponses = get_removed_keys_from_objects_1.getRemovedKeysFromObjects(sourceResponses, destinationResponses);
    return removedResponses.map((statusCode) => {
        const parsedResponse = sourceResponses[statusCode];
        return create_difference_1.createDifference({
            action: 'remove',
            destinationSpecOrigins: [],
            propertyName: 'response.status-code',
            source: 'openapi-diff',
            sourceSpecOrigins: [parsedResponse.originalValue]
        });
    });
};
exports.findDifferencesInResponses = (sourceResponses, destinationResponses) => {
    return [
        ...findAddedDifferencesInResponses(sourceResponses, destinationResponses),
        ...findRemovedDifferencesInResponses(sourceResponses, destinationResponses)
    ];
};
